keras_retinanet/bin/train.py

```diff
@@ -287,8 +254,37 @@ def csv_list(string):
 
     return check_args(parser.parse_args(args))
 
+def load_plugins(plugin_path):
+    """
+    Responsible for initialising the plugin manager, setting the plugin directory to search, and loading all available
+    plugins - then activating them.
+
+    :param plugin_path: String/[Str] for plugin paths to check
+    :return: None
+    """
+
+    pm = PluginManagerSingleton.get()
+    pl = pm.getPluginLocator()
+    pl.setPluginInfoExtension("dataset")
+    pm.setPluginLocator(pl)
+
+    plugin_path = [plugin_path] if type(plugin_path) is str else plugin_path
+    pm.setPluginPlaces(plugin_path)
+
+    # Load all plugins
+    pm.collectPlugins()
+
+    for k, n in enumerate(pm.getAllPlugins()):
+        pm.activatePluginByName(n.name)
+        print("Loaded: {}".format(n.name))
 
 def main(args=None):
+    #Load plugins first, as their procedures are needed for parsing args.
+    print("Loading plugins...")
+    # Load Plugins
+    load_plugins(['plugins'])
```

Comment:
I'd like to expose the plugin path to users. Ideally, the plugin path comes from the following places:
1. Command line flags.
2. An environment variable.
3. Some built-in default suitable for packaging. (That generally means a packager has to be able to set it during a configure step. Not sure how python programs normally deal with this.)

Ideally command line flags give enough control to precisely determine the the plugin path as desired. Default behaviour should probably be to append to the built-in path(s) and any paths from the environment, but allow disabling built-in paths and paths from the environment.

This isn't necessarily required before merging, but it's definitely something I want to have up to scratch. It would require changes to argument parsing to atleast a two-stage process.

Before merging though, I don't think it currently picks up the built-in plugins? It only looks in the `plugins` folder in the current working directory now, it seems.

Reply:
I believe the plugin folder should remain static, as part of RetinaNet. E.g Most games that utilise this pattern have a singular defined plugins folder, a central location for users to augment. I think if we start having multiple places to look for things become decentralised far too quickly. An organised location for encapsulating Datasets seems more sensible than a wild-west approach, personally. Thoughts?

Reply:
I think one big value of a plugin system is to allow third party contributions without modifying the source of the project. So allowing the loading of plugins from user-defined locations seems logical to me.

However, the current plugin path seems to be relative to the working directory. If we want to have a plugins folder in the keras-retinanet package, we should probably make it relative to the location of package?

Reply:
I'm not 100% with regard to the concreteness of paths in relation to the module. But typically all solutions which utilise Plugin systems have a designated folder for dropping 'modules'/plugins into. Adding more command line arguments will just complicate use of RetinaNet. A dedicated and singular 'bin' to drop things into reduces the amount of checking required. If we have 50 paths to check for plugins, that's a lot of folders, makes error checking harder.
The idea is that these plugins are designed to run with RetinaNet, so I don't see a need for them to be elsewhere really. If people need other directories as development areas, either copy them in, or symlink. Multiple folder locations certainly isn't the convention within industry from what I've seen; however, I'm happy to be proven wrong - I just don't see the added benefit.

Reply:
If the goal is to have one specific directory with additional plugins we don't even need something like yapsy. We could just have that directory be a python module and iterate over the submodules.

However, I always keep in mind the end-goal for any software project: having it packaged and installed system wide by a package manager (such as pacman). That means its files and directories will be root owned, and users will not be able to drop in any plugins.

Additionally, users may want to have different sets of plugins if they're unlucky enough that plugins they're interested in may have claimed conflicting names. 
That's why I think the user should have full control over the plugin path user by any application.

And to keep things easy for the user, there should be useful defaults that work most of the time.

But again, this is not crucial for the first step. But if we have a fixed path it must atleast always pick up the default bundled datasets.

Reply:
>If the goal is to have one specific directory with additional plugins we don't even need something like yapsy. We could just have that directory be a python module and iterate over the submodules.
As stated before, this was the first attempt for a plugin system; however, Yapsy handles a lot of error-handling and compatibility stuff. Especially as it's still a raw python implementation, we would simply be reinventing the wheel and it would take longer to put all the error checking in place.

>However, I always keep in mind the end-goal for any software project: having it packaged and installed system wide by a package manager (such as pacman). That means its files and directories will be root owned, and users will not be able to drop in any plugins.
I see your point. Indeed, if it is installed via a package manager and the plugins folder is owned by root, it may become more difficult for users to actually utilise the system. It then becomes important to augment plugins via some user-space locations.

I'd say flags, but we're getting to the stage where the incantations are huge. It's almost at a point where loading in a JSON/XML config file is more user friendly than args. I might lean towards environment variable use. It makes sense that they would define some region on their system for user level scripts. Packages can define environment variables when installing to the system, for the user. E.g. `RETINANET_PLUGIN_DIR` would be `~/.retinanet_plugins/` for example.

Reply:
The perfect solution for determining the plugin path to me is a combination of all of them: command line flags, an environment variable and some built-in default path(s) (in that order of priority). But an initial solution which only uses some built-in default(s) and an environment variable is also good. Even the environment bit is optional as far as I'm concerned for the first version.

If you do add an environment variable, I'd go for `RETINANET_PLUGIN_PATH` though, and allow colon-separated paths (like `$PATH`).